### PR TITLE
Allow reversing geodesic paths to the correct order

### PR DESCRIPTION
**Pull request by adeak**
Looking at https://github.com/pyvista/pyvista-support/issues/409 I realized that the path segments created by `enable_geodesic_picking` are each reversed within the given segment, which is the direct result of the `geodesic()` filter.

For example:
```py
>>> import pyvista as pv
>>> mesh = pv.Sphere()
>>> start, end = 100, 200
>>> geodesic = mesh.geodesic(start, end)
>>> geodesic.point_arrays['vtkOriginalPointIds']
pyvista_ndarray([200, 172, 144, 116,  88,  89,  90,  91,  92,  93,  94,
                  95,  96,  97,  98,  99, 100])
```
The path starting at point 100 and ending at point 200 is returned as `PolyData` whose first point is the end point and the last point is the start point:
```py
>>> import numpy as np
>>> np.array_equal(geodesic.points[0], mesh.points[start])
False
>>> np.array_equal(geodesic.points[0], mesh.points[end])
True
>>> np.array_equal(geodesic.points[-1], mesh.points[start])
True
```

This seems to be a direct consequence of the underlying VTK filter. Still, we can just (optionally) reverse the path, making it possible for geodesic point picking to give the correct path. And then we can enable the same behaviour for `Plotter.enable_geodesic_picking`. And when we enable to proper order for `enable_geodesic_picking` it also makes sense to clean the path, removing now-adjacent duplicate points.

I've also changed the code a bit so that the geodesic filter's `'vtkOriginalPointIds'` point array is preserved for the final geodesic path. This seems like something that can often be useful.

Since downstream users might have worked around the current behaviour of the geodesic filter I've decided to add this as opt-in. It would be nice if we could make it the default though.

### CODE REVIEW

**Comment by akaszynski** [comment]
> It would be nice if we could make it the default though.

I think making it default would be preferred.  This seems like a bug that `pyvista` is correcting, not modifying a feature of VTK.

File: pyvista/core/filters.py
Comment (akaszynski) [inline pyvista/core/filters.py:4444]:
```suggestion
                 keep_order=True):
```

Let's just keep this default.

**Comment by whophil** [comment]
Seems like something in this PR has considerably slowed down CI. The Linux "Test Package Docstring" step is taking an hour, see:
https://dev.azure.com/pyvista/PyVista/_build/results?buildId=6005&view=logs&jobId=b62462e1-cdfa-5405-7bf3-310b9a7831f1&j=b62462e1-cdfa-5405-7bf3-310b9a7831f1&t=ebc10e2c-4737-587e-d2e1-584909aaf71e

**Comment by adeak** [comment]
Thanks @akaszynski. I've [added a note](https://github.com/pyvista/pyvista/pull/1381/files#diff-6acd6a81eb79b8164e452c6b5e640e20732204f41d2ae092327b283c5acf31beR447-R451) in `enable_geodesic_picking()` to highlight the behaviour change, because I expect there might be downstream users who've been working around this bug (and their workaround will likely be broken by our fix).

I've also been wondering if we should add the version when this was changed, but we don't usually do that in pyvista so that would stick out.

**Comment by adeak** [comment]
I'm getting mixed messages here from CI. This page says all checks passed, but I'm getting emails pointing at failed CI in https://github.com/adeak/pyvista/actions/runs/935156102:
```
=================================== FAILURES ===================================
___________________ test_rotate_should_match_vtk_rotation[y] ___________________

axis = 'y'
grid = UnstructuredGrid (0x120688be0)
  N Cells:	40
  N Points:	99
  X Bounds:	0.000e+00, 1.000e+00
  Y Bounds:	0.000e+00, 1.000e+00
  Z Bounds:	0.000e+00, 5.000e+00
  N Arrays:	3


>   ???

tests/test_common.py:254: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/hypothesis/core.py:651: in execute_once
    self.__flaky(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <hypothesis.core.StateForActualGivenExecution object at 0x120805f40>
message = "Hypothesis test_rotate_should_match_vtk_rotation(angle=0.0, axis='y', grid=UnstructuredGrid (0x120688be0)\n  N Cells:...0, 5.000e+00\n  N Arrays:\t3) produces unreliable results: Falsified on the first call but did not on a subsequent one"

    def __flaky(self, message):
        if len(self.falsifying_examples) <= 1:
>           raise Flaky(message)
E           hypothesis.errors.Flaky: Hypothesis test_rotate_should_match_vtk_rotation(angle=0.0, axis='y', grid=UnstructuredGrid (0x120688be0)
E             N Cells:	40
E             N Points:	99
E             X Bounds:	0.000e+00, 1.000e+00
E             Y Bounds:	0.000e+00, 1.000e+00
E             Z Bounds:	0.000e+00, 5.000e+00
E             N Arrays:	3) produces unreliable results: Falsified on the first call but did not on a subsequent one

../../../hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/hypothesis/core.py:884: Flaky
---------------------------------- Hypothesis ----------------------------------
Falsifying example: test_rotate_should_match_vtk_rotation(
    angle=0.0, axis='y', grid=UnstructuredGrid (0x120688be0)
      N Cells:	40
      N Points:	99
      X Bounds:	0.000e+00, 1.000e+00
      Y Bounds:	0.000e+00, 1.000e+00
      Z Bounds:	0.000e+00, 5.000e+00
      N Arrays:	3,
)
Unreliable test timings! On an initial run, this test took 372.23ms, which exceeded the deadline of 200.00ms, but on a subsequent run it took 10.04 ms, which did not. If you expect this sort of variability in your test timings, consider turning deadlines off for this test by setting deadline=None.

You can reproduce this example by temporarily adding @reproduce_failure('6.14.0', b'AAAAAAUAAAAAAAAAAA==') as a decorator on your test case
```

Needless to say, I can't reproduce that error with the decorator.

**Comment by akaszynski** [comment]
Hypothesis drives me insane sometimes.  Not only does that test run long, but it gives me intermittent failures that can't be reproduced.

Would you be willing on figuring out how to clean that up?

**Comment by akaszynski** [comment]
> I've also been wondering if we should add the version when this was changed, but we don't usually do that in pyvista so that would stick out.

I see this in numpy as well.  I'm fine with adding a "added in version" note.  It would stick out, but someone's gotta be first.

**Comment by adeak** [comment]
> Would you be willing on figuring out how to clean that up?

I can take a look. There must be some edge cases where the test slows down if you keep seeing it, so there's hope.

> I see this in numpy as well. I'm fine with adding a "added in version" note. It would stick out, but someone's gotta be first.

OK, thanks. Then we only have to figure out if the next version will be a micro or minor version bump ;)

**Comment by akaszynski** [comment]
> OK, thanks. Then we only have to figure out if the next version will be a micro or minor version bump ;)

Put for the next minor version 0.32.0.  Any patch releases from here on out will be just patch versions.  We've fallen into a pattern of releasing patch releases for critical bugs, and feature (minor) releases every 4-8 weeks.  Not sure if that's too fast/slow.

**Review by akaszynski — APPROVED** [review]
It's up to you how to add the note regarding the version added.  I'm seeing a consistent pattern in `numpy` documentation.  See the note regarding the `like` kwarg.
https://numpy.org/doc/stable/reference/generated/numpy.zeros.html

**Comment by akaszynski** [comment]
Seems that we would run into issues if #1385 isn't merged first.  Let's wait until that's merged and then move our changes to the new filters files.

**Comment by akaszynski** [comment]
Looks like there's work to do due to #1385, which is why we went with #1385 first and then PRs that used filters second.

**Comment by adeak** [comment]
I've added the `.. versionadded` and `.. versionchanged` directives.

The former looks like this in the docs of `filters/poly_data.py`:
Screenshot 2021-06-19 at 13-07-04 pyvista PolyDataFilters geodesic — PyVista 0 31 dev0 documentation
(note the `keep_order` kwarg)

The latter looks like this in the docs of `plotting/picking.py`:
Screenshot 2021-06-19 at 13-07-04 pyvista PolyDataFilters geodesic — PyVista 0 31 dev0 documentation
(again note the `keep_order` kwarg)

I'm not crazy about the latter because it's a lot less obvious that there's something non-trivial there. Do you think people will read it anyway? But perhaps the original `.. note` was too loud anyway. I'm not sure about this either way.

**Comment by akaszynski** [comment]
> I'm not crazy about the latter because it's a lot less obvious that there's something non-trivial there

I prefer the `.. versionadded` directive.

**Comment by adeak** [comment]
How about note plus versionadded?
Screenshot 2021-06-19 at 22-07-42 pyvista Plotter enable_geodesic_picking — PyVista 0 31 dev0 documentation

**Comment by akaszynski** [comment]
That's even better.  These sorts of warnings really should be obvious.

**Comment by akaszynski** [comment]
Good to merge.

**Comment by adeak** [comment]
Thanks yet again for the feedback @akaszynski.

**Pull request merged**
